markovflow/kalman_filter.py

```diff
@@ -495,3 +494,119 @@ def _log_det_observation_precision(self):
     def observations(self):
         """ Observation vector """
         return self.sites.means
+
+
+@tf_scope_class_decorator
+class KalmanFilterWithSparseSites(BaseKalmanFilter):
+    r"""
+    Performs a Kalman filter on a :class:`~markovflow.state_space_model.StateSpaceModel`
+    and :class:`~markovflow.emission_model.EmissionModel`, with Gaussian sites, over a time grid.
+    """
+
+    def __init__(self, state_space_model: StateSpaceModel, emission_model: EmissionModel, sites: GaussianSites,
+                 time_grid_shape: tf.TensorShape, observations_index: tf.Tensor, observations: tf.Tensor):
+        """
+        :param state_space_model: Parameterises the latent chain.
+        :param emission_model: Maps the latent chain to the observations.
+        :param sites: Gaussian sites over the time grid.
+        :param time_grid_shape: Time grid shape, tuple (T, 1).
+        :param observations_index: Index of the observations in the time grid with shape (N,).
+        :param observations: Observations with shape (N, state_dim).
+        """
+        self.sites = sites
+        self.time_grid_shape = time_grid_shape
+        self.observations_index = observations_index
+        self._observations = tf.scatter_nd(self.observations_index, observations, self.time_grid_shape)
+        self.data_sites = self._get_data_sites()
+        super().__init__(state_space_model, emission_model)
+
+    def _get_data_sites(self):
+        """
+        Get data sites from all sites using observation idx.
+        """
+        nat1 = tf.gather_nd(self.sites.nat1, self.observations_index)
+        nat2 = tf.gather_nd(self.sites.nat2, self.observations_index)
+        log_norm = tf.gather_nd(self.sites.log_norm, self.observations_index)
+
+        return UnivariateGaussianSitesNat(
+            nat1=nat1,
+            nat2=nat2,
+            log_norm=log_norm,
+        )
+
+    @property
```

Comment:
```suggestion
     def sparse_to_dense(self, tensor):
          return tf.scatter_nd(tensor, self.observations_index)

     def dense_to_sparse(self, tensor):
          return tf.gather_nd(tensor, self.observations_index)


    @property
```

Reply:
these shouldn't work as is, but I think it would make the code more readable to add and use such functions throughout the code